src/box/tuple_compare.cc

```diff
@@ -232,34 +232,53 @@ mp_compare_decimal(const char *lhs, const char *rhs)
 
 }
 
+static inline int
+compare_decimal_double_infinite(double rhs)
+{
+	if (isfinite(rhs))
+		return 0;
+	/* We assume NaN is less than everything else. */
+	if (isnan(rhs))
+		return 1;
+	/* d is either +Inf or -Inf. */
+	return (rhs < 0) - (rhs > 0);
+}
+
 static int
 mp_compare_decimal_any_number(decimal_t *lhs, const char *rhs,
 			      enum mp_type rhs_type, int k)
 {
 	decimal_t rhs_dec;
+	decimal_t *rc;
 	switch (rhs_type) {
 	case MP_FLOAT:
 	{
 		double d = mp_decode_float(&rhs);
-		decimal_from_double(&rhs_dec, d);
+		int res = compare_decimal_double_infinite(d);
```

Comment:
Doing the special case handling before trying to convert the value to decimal means you check `isinf/isfinite` twice - first time in your function, second time in `decimal_from_double()`. It might be better to try the conversion first, and handle special cases only if it fails. See the diff (I didn't test it). But I don't mind your version too.
<details><summary>diff</summary>

```Diff
diff --git a/src/box/tuple_compare.cc b/src/box/tuple_compare.cc
index c387af4fa..0e58eba2e 100644
--- a/src/box/tuple_compare.cc
+++ b/src/box/tuple_compare.cc
@@ -232,15 +232,17 @@ mp_compare_decimal(const char *lhs, const char *rhs)
 
 }
 
+/**
+ * Compare decimal to 'not a number' which is either NaN or inf. Decimal value
+ * does not matter then. It is always a valid number.
+ */
 static inline int
-compare_decimal_double_infinite(double rhs)
+decimal_compare_nan(double rhs)
 {
-	if (isfinite(rhs))
-		return 0;
 	/* We assume NaN is less than everything else. */
 	if (isnan(rhs))
 		return 1;
-	/* d is either +Inf or -Inf. */
+	assert(isinf(rhs));
 	return (rhs < 0) - (rhs > 0);
 }
 
@@ -254,19 +256,17 @@ mp_compare_decimal_any_number(decimal_t *lhs, const char *rhs,
 	case MP_FLOAT:
 	{
 		double d = mp_decode_float(&rhs);
-		int res = compare_decimal_double_infinite(d);
-		if (res != 0)
-			return res * k;
 		rc = decimal_from_double(&rhs_dec, d);
+		if (rc == NULL)
+			return decimal_compare_nan(d) * k;
 		break;
 	}
 	case MP_DOUBLE:
 	{
 		double d = mp_decode_double(&rhs);
-		int res = compare_decimal_double_infinite(d);
-		if (res != 0)
-			return res * k;
 		rc = decimal_from_double(&rhs_dec, d);
+		if (rc == NULL)
+			return decimal_compare_nan(d) * k;
 		break;
 	}
 	case MP_INT:
```
</details>

Reply:
Thanks for the review and for the proposed changes!
I've applied your diff and force-pushed the updated patch.